File: Sources/UtilityC.m
#import <stdio.h>
#import <stdarg.h>
#import "UtilityC.h"

NSString* additionalSearchPath()
{
    NSString *resourcesDir = @"TeX2img.app/Contents/Resources";
    
    __block NSMutableArray<NSString*> *results = [NSMutableArray<NSString*> array];
    [GUI_PATHS enumerateObjectsUsingBlock:^(id  _Nonnull guiPath, NSUInteger idx, BOOL * _Nonnull stop) {
        [results addObject:[guiPath stringByAppendingPathComponent:[resourcesDir stringByAppendingPathComponent:@"mupdf"]]];
        [results addObject:[guiPath stringByAppendingPathComponent:[resourcesDir stringByAppendingPathComponent:@"pdftops"]]];
    }];
     
    return [results componentsJoinedByString:@":"];
}

void printStdErr(const char *format, ...)
{
    va_list list;
    va_start(list, format);
    vfprintf(stderr, format, list);
    va_end(list);
}

void suggestLatexOption()
{
    printStdErr("If you want to use another LaTeX compiler, specify it by using --latex option.\n");
}


BOOL checkWhich(NSString *cmdName)
{
    int status = system([NSString stringWithFormat:@"PATH=%@:$PATH; /usr/bin/which %@ > /dev/null", additionalSearchPath(), cmdName].UTF8String);
    return (status == 0) ? YES : NO;
}

NSString* getPath(NSString *cmdName)
{
    char str[PATH_MAX];
    FILE *fp;
    char *pStr;
    
    if ((fp = popen([NSString stringWithFormat:@"PATH=%@:$PATH; /usr/bin/which %@", additionalSearchPath(), cmdName].UTF8String, "r")) == NULL) {
        return nil;
    }
    fgets(str, PATH_MAX-1, fp);
    
    pStr = str;
    while ((*pStr != '\r') && (*pStr != '\n') && (*pStr != EOF)) {
        pStr++;
    }
    *pStr = '\0';
    
    if (pclose(fp) == 0) {
        return @(str);
    } else {
        return nil;
    }
}

File: Sources/Utility/UtilityC.swift
import Foundation

class UtilityC: NSObject {
    private static let whichPath = "/usr/bin/which"
    private static let resourcesDir = "TeX2img.app/Contents/Resources"

    private static func guiPaths() -> [String] {
        let home = NSHomeDirectory()
        return [
            home.appendingPathComponent("Desktop"),
            home.appendingPathComponent("Applications"),
            home,
            "/Applications",
            "/Applications/TeXLive",
        ]
    }

    private static func additionalSearchPath() -> String {
        var results = [String]()
        for guiPath in guiPaths() {
            let mupdfPath = guiPath.appendingPathComponent(resourcesDir.appendingPathComponent("mupdf"))
            let pdftopsPath = guiPath.appendingPathComponent(resourcesDir.appendingPathComponent("pdftops"))
            results.append(mupdfPath)
            results.append(pdftopsPath)
        }
        return results.joined(separator: ":")
    }

    private static func processEnvironment() -> [String: String] {
        var environment = ProcessInfo.processInfo.environment
        let additional = additionalSearchPath()
        let currentPath = environment["PATH"] ?? ""
        environment["PATH"] = "\(additional):\(currentPath)"
        return environment
    }

    static func printStdErr(_ message: String) {
        guard let data = message.data(using: .utf8) else { return }
        FileHandle.standardError.write(data)
    }

    static func suggestLatexOption() {
        printStdErr("If you want to use another LaTeX compiler, specify it by using --latex option.\n")
    }

    static func checkWhich(_ cmdName: String) -> Bool {
        let task = Process()
        task.executableURL = URL(fileURLWithPath: whichPath)
        task.arguments = [cmdName]
        task.environment = processEnvironment()
        task.standardOutput = FileHandle.nullDevice
        task.standardError = FileHandle.nullDevice
        try? task.run()
        task.waitUntilExit()
        return task.terminationStatus == 0
    }

    static func bundledToolPath(_ toolName: String, in subdirectory: String) -> String? {
        let executablePath = URL(fileURLWithPath: CommandLine.arguments[0]).standardizedFileURL
        let executableDirectory = executablePath.deletingLastPathComponent().path
        let candidates = [
            executableDirectory.appendingPathComponent(subdirectory).appendingPathComponent(toolName),
            executableDirectory.appendingPathComponent("../Resources/\(subdirectory)/\(toolName)"),
            executableDirectory.appendingPathComponent("../../Resources/\(subdirectory)/\(toolName)"),
            executableDirectory.appendingPathComponent("TeX2img.app/Contents/Resources/\(subdirectory)/\(toolName)"),
        ]

        for candidate in candidates {
            let standardized = candidate.standardizingPath
            if FileManager.default.isExecutableFile(atPath: standardized) {
                return standardized
            }
        }
        return nil
    }

    static func mudrawPath() -> String? {
        if let path = getPath("mudraw") {
            return path
        }
        return bundledToolPath("mudraw", in: "mupdf")
    }

    static func getPath(_ cmdName: String) -> String? {
        let task = Process()
        let pipe = Pipe()
        task.executableURL = URL(fileURLWithPath: whichPath)
        task.arguments = [cmdName]
        task.environment = processEnvironment()
        task.standardOutput = pipe
        task.standardError = FileHandle.nullDevice
        try? task.run()
        task.waitUntilExit()
        guard task.terminationStatus == 0 else { return nil }
        let output = pipe.stringValue.trimmingCharacters(in: .whitespacesAndNewlines)
        return output.isEmpty ? nil : output
    }
}